### ISSUE TITLE
FastAPI integration error

### ISSUE THREAD

**Issue by mcantrell** [pydantic/logfire#300]
### Description

I apologize if this belongs in the FastAPI issues instead of logfire. I'm not really sure who is the culprit here. 

I'm attaching a [sample project](https://github.com/user-attachments/files/16090524/example.zip) to demonstrate an error when `logfire[fastapi]` is added to a FastAPI project.

> **Note:** forcing a downgrade to pydantic v1 fixes the issue _(or removing logfire all together)_

### Error Reproduction

A simple API without any input works:

```shell
curl "http://localhost:8000/hello/"
```

Calling an API with a pydantic model as input fails:

```shell
curl -X POST "http://localhost:8000/test/" -H "Content-Type: application/json" -d '{"name": "test"}'
````

Error log:
```
INFO:     127.0.0.1:58776 - "POST /test/ HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 399, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/opentelemetry/instrumentation/asgi/__init__.py", line 631, in __call__
    await self.app(scope, otel_receive, otel_send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/routing.py", line 297, in handle
    await self.app(scope, receive, send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/routing.py", line 77, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/starlette/routing.py", line 72, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/fastapi/routing.py", line 269, in app
    solved_result = await solve_dependencies(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/logfire/_internal/integrations/fastapi.py", line 111, in patched_solve_dependencies
    return await instrumentation.solve_dependencies(request, original)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/logfire/_internal/integrations/fastapi.py", line 173, in solve_dependencies
    result = await original
             ^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/fastapi/dependencies/utils.py", line 628, in solve_dependencies
    ) = await request_body_to_args(  # body_params checked above
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/fastapi/dependencies/utils.py", line 758, in request_body_to_args
    v_, errors_ = field.validate(value, values, loc=loc)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/fastapi/_compat.py", line 127, in validate
    self._type_adapter.validate_python(value, from_attributes=True),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/pydantic/type_adapter.py", line 142, in wrapped
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/pydantic/type_adapter.py", line 373, in validate_python
    return self.validator.validate_python(object, strict=strict, from_attributes=from_attributes, context=context)
           ^^^^^^^^^^^^^^
  File "/Users/mcantrell/.pyenv/versions/3.11.9/lib/python3.11/functools.py", line 1001, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/pydantic/type_adapter.py", line 142, in wrapped
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/acme/fastapi-pydantic-error/.venv/lib/python3.11/site-packages/pydantic/type_adapter.py", line 318, in validator
    assert isinstance(self._validator, SchemaValidator)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

To fix the problem, either remove logfire completely or downgrade pydantic to v1.

### Python, Logfire & OS Versions, related packages (not required)

```TOML
requests="2.32.3"
pydantic="2.8.0"
fastapi="0.111.0"
protobuf="4.25.3"
rich="13.7.1"
executing="2.0.1"
opentelemetry-api="1.25.0"
opentelemetry-exporter-otlp-proto-common="1.25.0"
opentelemetry-exporter-otlp-proto-http="1.25.0"
opentelemetry-instrumentation="0.46b0"
opentelemetry-instrumentation-asgi="0.46b0"
opentelemetry-instrumentation-fastapi="0.46b0"
opentelemetry-proto="1.25.0"
opentelemetry-sdk="1.25.0"
opentelemetry-semantic-conventions="0.46b0"
opentelemetry-util-http="0.46b0"
```

**Comment by mcantrell** [comment]
[Sample Project](https://github.com/user-attachments/files/16090328/example.zip)

**Comment by alexmojaki** [comment]
@sydney-runkle this is a bug in type adapters in 2.8 when plugins are present.

Minimal repro:

```python
import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class TestModel(BaseModel):
    name: str


@app.post('/test/')
def test_model(model: TestModel):
    return model


uvicorn.run(app)
```

The error occurs when logfire is *installed* even if it's not used, because `self._validator` is a `PluggableSchemaValidator`.

@mcantrell you can work around this by disabling the pydantic plugin with the environment variable `PYDANTIC_DISABLE_PLUGINS=1`.

**Comment by mcantrell** [comment]
thanks for the  @alexmojaki

**Comment by alexmojaki** [comment]
This has been fixed in Pydantic 2.8.2
